### ISSUE TITLE
Install sos in an environment

### ISSUE THREAD

**Issue by avianinc** [vatlab/jupyterlab-sos#66]
Is there a way to install sos in a conda environment. I'm trying to install sos and elyra at the same time and it never works. I'm wondering if there is a way to install sos in an env. If I do get it to install right now... which happens occasionally depending on order of the installation I do not see the sos drop down combo box. I'm assuming this is jupyterlab 4.x related. I rolled the install back to 3.6.x but I never seems to install into the base environment correctly. Lastly, I'm building jupyterhub (not just lab).

**Comment by BoPeng** [comment]
Unfortunately, as reported by #65 as well, we do not yet support JupyterLab 4.x because it has some breaking API change that I currently do not have time to investigate. The only way to proceed is to use JupyterLab 3.x, which we have been using in a JupyterHub environment successfully.

I will see if I can have another look at JupyterLab 4.x this weekend.

**Comment by avianinc** [comment]
My real question should have been. Can I install sos in a conda environment under jupytehub. I have tried several ways but it never seems to work. I can see it in the notebook but not the lab. Do you have an example conda environment I can use to test the install? Thanks

**Comment by BoPeng** [comment]
I may have to consult our admin for this but as far as I remember, he installed jupyterhub using a conda environment, and then I used `pip install` to install sos related stuff to the conda environment on which jupyterhub/jupyterlab is running. I used `pip` instead of `conda` for installation because the `pip` versions are more up to date. (e.g. the latest version of `sos-r` is not available from conda-forge).

**Comment by BoPeng** [comment]
JLab 4.x is now (partially) supported.

**Comment by avianinc** [comment]
Bo... thanks for the update. I'm trying my best to install SoS. It's not happening anymore. I have spent hours trying to install this on windows, linux, using both conda and pip methods. I cannot get any working environments. Would you happen to have a working environment you could share? Regards,

**Comment by BoPeng** [comment]
Please be more specific about what you have done, which versions of packages you have installed so that I can better assist you. But in general,

1. you can use the sos docker image (https://vatlab.github.io/sos-docs/running.html#content) if you can use docker. It is not updated for JLab 4.x but it works.
2. You need to use a new conda environment so that I can reproduce your problem. The installation is actually pretty easy usually. WIthout checking syntax etc, 

```
conda create --name jlab python=3.10 -y
cona activate jlab
pip install jupyterlab jupyterlab-sos sos-python
python -m sos_notebook.install
```

If things do not work, send me the output of

```
pip list | grep jupyter
pip list | grep sos
jupyter kernelspec list
jupyter labextension list
```

**Comment by avianinc** [comment]
Bo... thanks for getting back to me.. Here is my output. Please take a look and let me know if you see anything. I have tries this across several os.

(jlab) ***@***.***:/# pip list | grep jupyter
jupyter_client            8.3.1
jupyter_core              5.3.2
jupyter-events            0.7.0
jupyter-lsp               2.2.0
jupyter_server            2.7.3
jupyter_server_terminals  0.4.4
jupyterlab                4.0.6
jupyterlab-pygments       0.2.2
jupyterlab_server         2.25.0
jupyterlab_sos            0.10.0

(jlab) ***@***.***:/# pip list | grep sos
jupyterlab_sos            0.10.0
sos                       0.24.3
sos-notebook              0.24.4
sos-python                0.20.0

(jlab) ***@***.***:/# jupyter kernelspec list
Available kernels:
  python3    /opt/conda/envs/jlab/share/jupyter/kernels/python3
  sos        /usr/local/share/jupyter/kernels/sos

(jlab) ***@***.***:/# jupyter labextension list
JupyterLab v4.0.6
/opt/conda/envs/jlab/share/jupyter/labextensions
        jupyterlab_pygments v0.2.2 enabled  X (python, jupyterlab_pygments)
        jupyterlab-sos v0.8.9 enabled OK (python, jupyterlab_sos)
        transient-display-data v0.4.4 enabled OK (python, transient-display-data)
        @jupyter-notebook/lab-extension v7.0.4 enabled OK


   The following extensions are outdated:
        jupyterlab_pygments

   Consider checking if an update is available for these packages.

Here is a snapshot of what Im seeing when I use the %get from python. I hope I'm not missing something but the dependencies never seem to work out.

***@***.***

Thanks for your help,

From: Bo ***@***.***>
Sent: Friday, October 6, 2023 1:06 PM
To: vatlab/jupyterlab-sos ***@***.***>
Cc: John DeHart ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL]Re: [vatlab/jupyterlab-sos] Install sos in an environment (Issue #66)

CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.


Please be more specific about what you have done, which versions of packages you have installed so that I can better assist you. But in general,

  1.  you can use the sos docker image (https://vatlab.github.io/sos-docs/running.html#content<https://usg02.safelinks.protection.office365.us/?url=https%3A%2F%2Fvatlab.github.io%2Fsos-docs%2Frunning.html%23content&data=05%7C01%7Cjdehart%40avian.com%7Caeeda724aaf048ca366b08dbc68e795f%7C8965f906c4b04dd1b47d0e6c58d820c1%7C0%7C0%7C638322087470566554%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=38vHgtBd%2F717rOb0oevnmQ8O9FAsm7e%2B9iDTJEEfWOg%3D&reserved=0>) if you can use docker. It is not updated for JLab 4.x but it works.
  2.  You need to use a new conda environment so that I can reproduce your problem. The installation is actually pretty easy usually. WIthout checking syntax etc,

conda create --name jlab python=3.10 -y

cona activate jlab

pip install jupyterlab jupyterlab-sos sos-python

python -m sos_notebook.install

If things do not work, send me the output of

pip list | grep jupyter

pip list | grep sos

jupyter kernelspec list

jupyter labextension list

-

**Comment by BoPeng** [comment]
First the version of `jupyterlab-sos ` does not look right, pip says 0.10.0, but labextension says 0.8.9 . Then, do you see the right hand side kernel selector from jupyterlab? I do not see it from your screenshot. The output of `%use` from jupyter will also show status of available kernels.

**Comment by avianinc** [comment]
Bo... I tried the docker image and the same thing is happening. See the included image. This is from a fresh build of the docker image on Linux. You can see that the python kernel is not able to '%get' but in the next cell you can see that R works fine. 

![image](https://github.com/vatlab/jupyterlab-sos/assets/93828723/cf3bf9cb-1f0f-4cb4-9b54-23f35afacba7)

So I'm not really interested in Jupyter 4.x, I just need a stable build out that works. So there are two things:

1. I need to make sure that its just not me that is having this issue with the python kernel. I have this issue no matter the OS or install method.
2. Can you share the docker file you use for the build? I would like to use it as a base for my next build operations.

Thanks,

**Comment by BoPeng** [comment]
On mac,

<img width="1517" alt="image" src="https://github.com/vatlab/jupyterlab-sos/assets/9889312/256fd55c-560c-4c26-9042-115ba3100e50">
<img width="1178" alt="image" src="https://github.com/vatlab/jupyterlab-sos/assets/9889312/2589d0a9-281b-4df8-ac10-a9e23725ddbd">

**Comment by avianinc** [comment]
Bo... I have tried this everywhere today, aws, linux, docker, no docker, and windows (I don't have a mac...). All of the kernels I have installed work but python. I'm using Jupyterlab 3.6.x due to an incompatibility with another tool I must use. Also, I cannot get the SoS dropdowns to show no matter what combination of JL4.X I use. Any chance you might ask someone to run this on a linux or windows box to verify what I'm seeing or even try the sos docker image. This has been an issue for over a year, I kept thinking it was my installation causing the problem (still could be...) but tried conda, pip, virtualenv, basic jupyterlab installs on several linux flavors. _The kicker for me was pulling the sos docker image with the same result right out of the build._ 

![image](https://github.com/vatlab/jupyterlab-sos/assets/93828723/2907905c-899f-4f93-9f31-7bedd1c6ace9)
![image](https://github.com/vatlab/jupyterlab-sos/assets/93828723/eaf463d1-05c1-42b3-803b-597db9ab87ac)
![image](https://github.com/vatlab/jupyterlab-sos/assets/93828723/28b6c98e-ed10-4301-b6f4-54e8013ae818)

Thanks for all your help here, I really like SoS and wanted to start up again but I'm stuck....

**Comment by BoPeng** [comment]
Thank you for your patience. There are many moving pieces with jupyter and sos-notebook but I can now reproduce the problem with the official docker image. Let me check what is going on.

**Comment by BoPeng** [comment]
This turns out to be an issue of `sos-python` with python >= 3.10. Please update `sos-python` to 0.20.1 and see if the problem persists.

**Comment by avianinc** [comment]
YES!!!! That was a tough one. Thank you for sticking with me here. The python integration is working now on windows and linux...
